proposals/records.md

```diff
@@ -1,54 +1,60 @@
     ;
 
-struct_declaration
-    : attributes? struct_modifiers? 'partial'? 'struct' identifier type_parameter_list?
-      parameter_list? struct_interfaces? type_parameter_constraints_clauses? struct_body
+record_base
+    : ':' class_type argument_list?
+    | ':' interface_type_list
+    | ':' class_type argument_list? interface_type_list
     ;
 
-class_body
-    : '{' class_member_declarations? '}'
-    | ';'
-    ;
-
-struct_body
-    : '{' struct_members_declarations? '}'
+record_body
+    : '{' class_member_declaration* '}'
     | ';'
     ;
 ```
 
-The `attributes` non-terminal will also permit a new contextual attribute, `data`.
-
-A class (struct) declared with a parameter list or `data` modifier is called a record class (record struct), either of which is a record type.
-
-It is an error to declare a record type without the `data` modifier.
+Record types are reference types, similar to a class declaration. It is an error for a record to provide
+a `record_base` `argument_list` if the `record_declaration` does not contain a `parameter_list`.
 
 ## Members of a record type
 
-In addition to the members declared in the class or struct body, a record type has the following additional members:
+In addition to the members declared in the record body, a record type has additional synthesized members.
+Members are synthesized unless an accessible concrete (non-abstract) member with a "matching" signature is 
+either inherited or declared in the record body. Two members are considered matching if they have the same
+signature or would be considered "hiding" in an inheritance scenario.
+
+The synthesized members are as follows:
 
 ### Equality members
 
-Record types produce synthesized implementations for the following methods:
+Record types produce synthesized implementations for the following methods, where `T` is the
+containing type:
 
-* `object.GetHashCode()` override, unless it is sealed or user provided
-* `object.Equals(object)` override, unless it is sealed or user provided
+* `object.GetHashCode()` override
+* `object.Equals(object)` override
 * `T Equals(T)` method, where `T` is the current type
+* `Type EqualityContract` get-only property
 
-`T Equals(T)` is specified to perform value equality such that `Equals` is
-true if and only if all the instance fields declared in the receiver type
-are equal to the fields of the other type.
+If either `object.GetHashCode()` or `object.Equals(object)` are sealed, an error is produced.
+
+`EqualityContract` is a virtual instance property which returns `typeof(T)`. If the base type
+defines an `EqualityContract` it is overridden in the derived record. If the base `EqualityContract`
+is sealed or non-virtual, an error is produced.
+
+`T Equals(T)` is specified to perform value equality such that `Equals` is true if and only if
+all accessible instance fields in the receiver are equal to the fields of the parameter
+and `this.EqualityContract` equals `other.EqualityContract`.
```

Comment:
The base class `Equals(T)` will ignore derived class fields Is that expected?

For instance, the following will print: `False, True`:
```C#
using System;

data class A(int X);

data class B(int X, int Y) : A(X);

class Program
{
    static void Main()
    {
        var b1 = new B(1, 2);
        var b2 = new B(1, 3);
        Console.WriteLine(b1.Equals(b2));
        Console.WriteLine(((A)b1).Equals(b2));
    }
}
```

Reply:
There will be different `EqualityContract` values for `A` & `B` - so even if you are calling `A.Equals()`, it can detect that they are not the same type.